src/rhsmlib/services/entitlement.py

```diff
@@ -0,0 +1,128 @@
+            return {'status': overall_status, 'reasons': reasons, 'valid': valid}
+        else:
+            return {'status': 'Unknown', 'reasons': {}, 'valid': False}
+
+    def get_pools(self, pool_subsets=None, matches=None, pool_only=None, match_installed=None,
+            no_overlap=None, service_level=None, show_all=None, on_date=None, **kwargs):
+        # We accept a **kwargs argument so that the DBus object can pass whatever dictionary it receives
+        # via keyword expansion.
+        if kwargs:
+            raise exceptions.ValidationError(_("Unknown arguments: %s") % kwargs.keys())
+
+        # [] or None means look at all pools
+        if not pool_subsets:
+            pool_subsets = ['installed', 'consumed', 'available']
+
+        options = {
+            'pool_subsets': pool_subsets,
+            'matches': matches,
+            'pool_only': pool_only,
+            'match_installed': match_installed,
+            'no_overlap': no_overlap,
+            'service_level': service_level,
+            'show_all': show_all,
+            'on_date': on_date,
+        }
+        self.validate_options(options)
+        results = {}
+        if 'installed' in pool_subsets:
+            installed = products.InstalledProducts(self.cp).list(options['matches'])
+            results['installed'] = [x._asdict() for x in installed]
+        if 'consumed' in pool_subsets:
+            results['consumed'] = self.get_consumed_product_pools(**options)
+        if 'available' in pool_subsets:
+            results['available'] = self.get_available_pools(**options)
+
+        return results
+
+    def get_consumed_product_pools(self, service_level=None, matches=None, **kwargs):
+        certs = self.entitlement_dir.list()
+        cert_filter = utils.EntitlementCertificateFilter(filter_string=matches, service_level=service_level)
+
+        if service_level is not None or matches is not None:
+            certs = list(filter(cert_filter.match, certs))
+
+        return certs
+
+    def get_available_pools(self, show_all=None, on_date=None, no_overlap=None,
+            match_installed=None, matches=None, service_level=None, **kwargs):
+        available_pools = managerlib.get_available_entitlements(
+            get_all=show_all,
+            active_on=on_date,
+            overlapping=no_overlap,
+            uninstalled=match_installed,
+            filter_string=matches
+        )
+
+        def filter_pool_by_service_level(pool_data):
+            pool_level = ""
+            if pool_data['service_level']:
+                pool_level = pool_data['service_level']
+            return service_level.lower() == pool_level.lower()
+
+        if service_level is not None:
+            available_pools = list(filter(filter_pool_by_service_level, available_pools))
+
+        return available_pools
+
+    def validate_options(self, options):
+        if not set(['installed', 'consumed', 'available']).issuperset(options['pool_subsets']):
```

Comment:
And the type error is handled in 130b570.  Thanks for catching that one.  The service was returning an EntitlementCertificate object rather than something that could be serialized to JSON.